File: java/src/leetcode/editor/cn/94-BinaryTreeInorderTraversal.java
package leetcode.editor.cn;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import ivonhoe.java.leetcode.bean.TreeNode;

/*
 * @lc app=leetcode.cn id=94 lang=java
 *
 * [94] Binary Tree Inorder Traversal
 *
 * https://leetcode-cn.com/problems/binary-tree-inorder-traversal/description/
 *
 * algorithms
 * Easy (75.66%)
 * Likes:    1638
 * Dislikes: 0
 * Total Accepted:    1M
 * Total Submissions: 1.4M
 * Testcase Example:  '[1,null,2,3]'
 *
 * Given the root of a binary tree, return the inorder traversal of its nodes'
 * values.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: root = [1,null,2,3]
 * Output: [1,3,2]
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: root = []
 * Output: []
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: root = [1]
 * Output: [1]
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The number of nodes in the tree is in the range [0, 100].
 * -100 <= Node.val <= 100
 * 
 * 
 * 
 * Follow up: Recursive solution is trivial, could you do it iteratively?
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        return inorderIteration(root);
    }

    private List<Integer> inorderIteration(TreeNode root) {
        List<Integer> ans = new ArrayList<Integer>();
        if (root == null) {
            return ans;
        }
        // 需要经过两次出栈
        Stack<Object> stack = new Stack<Object>();
        stack.push(root);
        while (!stack.isEmpty()) {
            Object node = stack.pop();
            if (node instanceof TreeNode) {
                TreeNode treeNode = (TreeNode) node;
                if (treeNode.right != null) {
                    stack.push(treeNode.right);
                }
                stack.push(treeNode.val);
                if (treeNode.left != null) {
                    stack.push(treeNode.left);
                }
            } else {
                ans.add((Integer) node);
            }
        }

        return ans;
    }

    private void inorder(TreeNode root, List<Integer> ans) {
        if (root == null) {
            return;
        }

        inorder(root.left, ans);
        ans.add(root.val);
        inorder(root.right, ans);
    }
}
// @lc code=end

File: kotlin/src/leetcode/editor/cn/94-BinaryTreeInorderTraversal.kt
//给定一个二叉树的根节点 root ，返回它的 中序 遍历。 
//
// 
//
// 示例 1： 
//
// 
//输入：root = [1,null,2,3]
//输出：[1,3,2]
// 
//
// 示例 2： 
//
// 
//输入：root = []
//输出：[]
// 
//
// 示例 3： 
//
// 
//输入：root = [1]
//输出：[1]
// 
//
// 示例 4： 
//
// 
//输入：root = [1,2]
//输出：[2,1]
// 
//
// 示例 5： 
//
// 
//输入：root = [1,null,2]
//输出：[1,2]
// 
//
// 
//
// 提示： 
//
// 
// 树中节点数目在范围 [0, 100] 内 
// -100 <= Node.val <= 100 
// 
//
// 
//
// 进阶: 递归算法很简单，你可以通过迭代算法完成吗？ 
// Related Topics 栈 树 哈希表 
// 👍 846 👎 0

import leetcode.editor.bean.TreeNode
import java.util.*
import kotlin.collections.ArrayList

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    /**
     * 中序遍历使用递归或dfs
     */
    fun inorderTraversal(root: TreeNode?): List<Int> {
        return solution2(root)
    }

    // dfs遍历
    fun solution2(root: TreeNode?): List<Int> {
        var result = ArrayList<Int>()
        var stack = LinkedList<TreeNode>()

        var cur = root
        while (!stack.isEmpty() || cur != null) {
            if (cur != null) {
                stack.push(cur)
                cur = cur.left
            } else {
                var nod = stack.pop()
                result.add(nod.`val`)
                cur = nod.right
            }
        }

        return result
    }

    // 递归
    fun solution1(root: TreeNode?): List<Int> {
        var result = ArrayList<Int>()

        fun inOrder(root: TreeNode?) {
            if (root == null) {
                return
            }
            inOrder(root.left)
            result.add(root.`val`)
            inOrder(root.right)
        }

        inOrder(root)
        return result
    }
}
//leetcode submit region end(Prohibit modification and deletion)
